### PR TITLE
Fix parsing for arrays of values in image changes

### PR DESCRIPTION
**Pull request by mheon**
The current parser interprets ["echo"] as the literal string "echo" - quotes and all. The correct behavior per Docker is to treat it as a JSON string - that is, the quotes delineate the beginning and end of the string, and are NOT included.

### CODE REVIEW

**Comment by mheon** [comment]
@baude PTAL - should fix your quoting issue

**Comment by mheon** [comment]
Test failure is consistent, I'll check it tomorrow morning.

File: pkg/util/utils.go [hunk 1]
```diff
@@ -79,18 +80,20 @@ func ParseChanges(option string) (key string, vals []string, err error) {
...
 		if len(tokens) < 2 {
 			return "", []string{}, fmt.Errorf("invalid key value %s", option)
 		}
 		key = strings.Trim(tokens[0], " ") // Need to trim whitespace part of delimiter.
-		val = tokens[1]
-		if strings.Contains(tokens[1], "[") && strings.Contains(tokens[1], "]") {
-			//Trim '[',']' if exist.
-			val = strings.TrimLeft(strings.TrimRight(tokens[1], "]"), "[")
+
+		if strings.Contains(tokens[1], "[") || strings.Contains(tokens[1], "]") {
```
Comment (TomSweeneyRedHat) [inline pkg/util/utils.go]: probably doesn't matter, but should this be an "and" rather than an "or"?

File: pkg/util/utils.go [hunk 2]
```diff
@@ -79,18 +80,20 @@ func ParseChanges(option string) (key string, vals []string, err error) {
...
+			// Try a JSON unmarshal to catch case 3
+			if err := json.Unmarshal([]byte(tokens[1]), &vals); err != nil {
+				return "", []string{}, errors.Wrapf(err, "cannot unmarshal %s as JSON array", tokens[1])
```
Comment (TomSweeneyRedHat) [inline pkg/util/utils.go]: should we fail out here if the umarshall fails or just stuff tokens[1] into vals and continue on after noting the unmarshal failure in a debug message?

Comment (mheon) [reply pkg/util/utils.go]: I was debating that, but the tests fail out on those situations, so long as `[` or `]` are present, so I kept that behaviour for now.

File: pkg/util/utils.go (hunk 1 above)
Comment (mheon) [reply pkg/util/utils.go]: An or - so we can detect invalid JSON arrays like `["hello", "world"` (missing `]`) and similar

**Comment by baude** [comment]
tested, LGTM

**Comment by mheon** [comment]
I *think* we're going to go green now, @baude

**Comment by mheon** [comment]
Pushed major changes - ended up needing a near-complete rewrite of parsing for `--change` to try and restore some sanity.

File: pkg/util/utils.go [hunk 3]
```diff
@@ -71,118 +73,175 @@ func StringInSlice(s string, sl []string) bool {
...
+
+	config := v1.ImageConfig{}
+
+	for _, change := range changes {
+		// First, let's assume proper Dockerfile format - space
+		// separator between instruction and value
+		split := strings.SplitN(change, " ", 2)
+
+		if len(split) != 2 {
+			split = strings.SplitN(change, "=", 2)
```
Comment (mheon) [inline pkg/util/utils.go]: I'm debating dropping this... It'll break out test suite and anyone who used our old format, but it also makes parsing so much harder than it needs to be.

**Comment by mheon** [comment]
Still working on tests

**Comment by mheon** [comment]
Tests are done

**Comment by mheon** [comment]
Finally green.

@baude @rhatdan @vrothberg @giuseppe  PTAL

File: pkg/util/utils.go [hunk 4]
```diff
@@ -71,118 +73,221 @@ func StringInSlice(s string, sl []string) bool {
...
-				labels[vals[0]] = vals[1]
-			} else {
-				labels[vals[0]] = ""
+			// Same general idea as ENV, but we no longer allow " "
+			// as a separator.
+			// We didn't do that for ENV either, so nice and easy.
+			// Potentially problematic: LABEL might theoretically
+			// allow an = in the key? If people really do this, we
+			// may need to investigate more advanced parsing.
+			var (
```
Comment (TomSweeneyRedHat) [inline pkg/util/utils.go]: Possible suggestion for later.  It looks like you're doing this key/vall parsing in at least two places.  You might consider moving it to a small function.  No change needed now, just a thought.

**Comment by mheon** [comment]
Apparently we have a completely separate set of parsing code in libpod's `Commit()` function. I'm going to tear that out and convert to the standard change parsing.

**Comment by mheon** [comment]
Pushed a new commit that swaps `podman commit` over to the new parsing logic.

**Comment by mheon** [comment]
Still to-do: re-add ONBUILD support

**Comment by mheon** [comment]
ONBUILD support re-added

**Comment by mheon** [comment]
It's green again. @baude @rhatdan @vrothberg PTAL and merge

**Pull request merged**